app/models/concerns/path_parser.rb

```diff
@@ -1,63 +1,46 @@
-    end
-
-    unless tags.empty?
-      tag_list.add(remove_stop_words(tags))
+  def parse_metadata_from_path!
+    if SiteSettings.model_path_template
+      components = extract_path_components
+      if components[:tags].present?
+        tag_list.add(remove_stop_words(components[:tags]))
+      end
+      self.creator = Creator.find_or_create_by(name: components[:creator]) if components[:creator]
+      collection_list.add(components[:collection]) if components[:collection]
       save!
     end
   end
+end
 
-  def autogenerate_creator_from_prefix_template!
-    if SiteSettings.model_tags_tag_model_path_prefix
-      filepaths = File.dirname(path).split("/")
-      filepaths.shift
-      templatechunks = SiteSettings.model_path_prefix_template.split("/")
-      creatornew = ""
-      collectionnew = ""
-      tags = []
-      while !templatechunks.empty? && !filepaths.empty? && !(templatechunks.length == 1 && templatechunks[0] == "{tags}")
-        if templatechunks[0] == "{creator}"
-          creatornew = filepaths.shift
-          templatechunks.shift
-        elsif templatechunks[0] == "{collection}"
-          collectionnew = filepaths.shift
-          templatechunks.shift
-        elsif templatechunks[-1] == "{creator}"
-          creatornew = filepaths.pop
-          templatechunks.pop
-        elsif templatechunks[-1] == "{collection}"
-          collectionnew = filepaths.pop
-          templatechunks.pop
-        else
-          logger.error "Invalid model_path_prefix_template: #{SiteSettings.model_path_prefix_template} / #{templatechunks[0]}"
-          templatechunks.shift
-        end
-      end
-      if templatechunks.length == 1 && templatechunks[0] == "{tags}"
-        tags = filepaths
-      end
-      unless tags.empty?
-        tag_list.add(remove_stop_words(tags))
+def path_parse_pattern
+  Regexp.new("^/?.*?" +
+    SiteSettings.model_path_template.gsub(/{.+?}/) { |token|
+      case token
+      when "{tags}"
+        "(?<tags>[[:print:]]*)"
+      when "{creator}"
+        "(?<creator>[[:print:]&&[^/]]*?)"
+      when "{collection}"
+        "(?<collection>[[:print:]&&[^/]]*?)"
+      when "{modelName}"
+        "(?<model_name>[[:print:]&&[^/]]*?)"
+      when "{modelId}"
+        "(?<model_id>#[[:digit:]]+)?"
+      else
+        "[[:print:]&&[^/]]*"
       end
-      unless creatornew.empty?
-        creator = Creator.find_by(name: creatornew)
-        creator ||= Creator.create(name: creatornew)
-        self.creator_id = creator.id
-      end
-      unless collectionnew.empty? && !collection_list
-        collection_list.add(collectionnew)
-      end
-      save!
-    end
-  end
+    } + "$")
+end
+
+def extract_path_components
+  components = path.match(path_parse_pattern)&.named_captures&.symbolize_keys
+  return {} if components.nil?
+  components.merge({
+    tags: components[:tags]&.split("/")&.compact_blank,
+    model_id: nil # discard ID, never gonna use it in parsing
```

Comment:
This *could* be used as a basis to figure out when someone did local file changes.  That might help avoid duplication when someone adjusts stuff manually.

Reply:
true, if the ID exists, we could use it to match if the folder path changes.

Reply:
one for the backlog I think

Reply:
agreed.